Assets/Scripts/UI/GalleryMenu.cs

```diff
@@ -0,0 +1,221 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.UI;
+using TMPro;
+using UnityEngine.InputSystem;
+
+[System.Serializable]
+public struct Weapon
+{
+    public Item body;
+    public Item barrel;
+    public Item extension;
+
+    public Weapon(Item Body, Item Barrel, Item Extension)
+    {
+        body = Body;
+        barrel = Barrel;
+        extension = Extension;
+    }
+    public Weapon(Item Body, Item Barrel)
+    {
+        body = Body;
+        barrel = Barrel;
+        extension = null;
+    }
+}
+public class GalleryMenu : MonoBehaviour
+{
+    [SerializeField]
+    private Weapon[] unlockedElements;
+    private int pageIndex = 0;
+    private int maxPages = 0;
+    private List<GameObject> spawnedWeapons = new List<GameObject>();
+    private List<RectTransform> gridElements = new List<RectTransform>();
+
+    [Header("Dependencies")]
+    public Transform gridBase;
+    public Transform navigation;
+    public TabsButton tabPrefab;
+
+    [Header("Tabs")]
+    public Color disabledColor;
+    public Color selectedColor;
+    private TabsButton selectedTab;
+    private List<TabsButton> tabs = new List<TabsButton>();
+
+    private MainMenuController mainMenuController;
+    private void Start()
+    {
+        unlockedElements = CreateAllWeapons();
+        mainMenuController= GetComponentInParent<MainMenuController>();
+
+        gridElements = gridBase.GetComponentsInChildren<Image>(true).Select(x => x.GetComponent<RectTransform>()).ToList();
+        gridElements.RemoveAt(0); // GetComponentInChildren returns this element as well, which we don't want
+
+        maxPages = Mathf.CeilToInt((float)unlockedElements.Length / gridElements.Count);
+
+        CreateTabs(maxPages);
+        PopulateGrid(0);
+    }
+
+    public void SetPlayerInput(InputManager inputManager)
+    {
+        inputManager.onLeftTab += PrevPage;
+        inputManager.onRightTab += NextPage;
+        inputManager.onCancel += Back;
+    }
+
+    private void CreateTabs(int pages)
+    {
+        for (int i = 1; i <= pages; i++)
+        {
+            TabsButton tab = Instantiate(tabPrefab, navigation.transform).GetComponent<TabsButton>();
+            tab.GetComponentInChildren<TMP_Text>().text = (i).ToString();
+            tabs.Add(tab);
+            tab.GetComponent<Image>().color = disabledColor;
+        }
+
+        selectedTab = tabs[0];
+        selectedTab.GetComponent<Image>().color = selectedColor;
+    }
+
+    /// <summary>
+    /// Update the grid items with the corresponding page of items
+    /// </summary>
+    void PopulateGrid(int page)
+    {
+        // Remove previous guns
+        foreach (GameObject g in spawnedWeapons)
+        {
+            GameObject.DestroyImmediate(g);
```

Comment:
`DestroyImmediate` seems dangerous according to [Unity's documentation](https://docs.unity3d.com/ScriptReference/Object.DestroyImmediate.html), could we get away with using `Destroy` here or is there a reason you use the _immediate_ version?

Reply:
changed